osu.Game.Tests/NonVisual/OverridableBindableTest.cs

```diff
@@ -0,0 +1,153 @@
+            setting.HasCustomValue.Value = true;
+
+            Assert.That(setting.FinalValue.Value, Is.EqualTo(8));
+            Assert.That(setting.FinalValue.Disabled, Is.False);
+        }
+
+        [Test]
+        public void TestCustomExpectedParsing()
+        {
+            var setting = new OverridableBindable<float>(
+                minValue: 2,
+                maxValue: 8,
+                defaultValue: 2);
+
+            Assert.That(setting.HasCustomValue.Value, Is.False);
+            Assert.That(setting.FinalValue.Value, Is.EqualTo(2));
+
+            parseAndEnsureExpected(7f, 7f);
+            parseAndEnsureExpected(new Bindable<float>(6f), 6f);
+            parseAndEnsureExpected(new OverridableBindable<float>(5f) { HasCustomValue = { Value = true } }, 5f);
+
+            parseAndEnsureExpected(14f, 8f);
+            parseAndEnsureExpected(-5f, 2f);
+
+            void parseAndEnsureExpected(object input, object expected)
+            {
+                setting.HasCustomValue.Value = false;
+                setting.Parse(input);
+
+                Assert.That(setting.HasCustomValue.Value, Is.True);
+                Assert.That(setting.FinalValue.Value, Is.EqualTo(setting.CustomValue.Value));
+                Assert.That(setting.FinalValue.Value, Is.EqualTo(expected));
+            }
+        }
+
+        [Test]
+        public void TestBaseExpectedParsing()
+        {
+            var setting = new OverridableBindable<float>(
+                minValue: 0,
+                maxValue: 10,
+                defaultValue: 5);
+
+            parseAndEnsureSame(new OverridableBindable<float>(9f));
+            parseAndEnsureSame(null);
+
+            void parseAndEnsureSame(object input)
+            {
+                setting.HasCustomValue.Value = true;
+                setting.Parse(input);
+
+                Assert.That(setting.HasCustomValue.Value, Is.False);
+                Assert.That(setting.FinalValue.Value, Is.EqualTo(setting.BaseValue.Value));
+                Assert.That(setting.FinalValue.Value, Is.EqualTo(5f));
+
+                // custom value shouldn't be touched as well.
+                Assert.That(setting.CustomValue.Value, Is.EqualTo(5f));
+            }
+        }
+
+        [Test]
+        public void TestSerialization()
+        {
+            var setting = new OverridableBindable<float>(
+                minValue: 0,
+                maxValue: 10,
+                defaultValue: 5);
+
+            setting.HasCustomValue.Value = true;
+
+            Assert.That(setting.Serialize().Deserialize<float>(), Is.EqualTo(5f));
+
+            var deserializedSetting = setting.Serialize().Deserialize<OverridableBindable<float>>();
+            Assert.That(deserializedSetting.HasCustomValue.Value, Is.True);
+            Assert.That(deserializedSetting.CustomValue.Value, Is.EqualTo(5f));
+
+            setting.HasCustomValue.Value = false;
+
+            Assert.That(setting.Serialize().Deserialize<float?>(), Is.Null);
```

Comment:
should have equivalent testing to the `HasCustomValue = true` case (deserialising to an `OverridableBindable<float>` should yield a non-null one with `HasCustomValue = true`)

Reply:
that is doing so right above this code

Reply:
i'm not seeing a

```csharp
            var deserializedSetting = setting.Serialize().Deserialize<OverridableBindable<float>>();
            Assert.That(deserializedSetting.HasCustomValue.Value, Is.False);
```

anywhere? i was explicitly saying that the deserialise/parse should also cover the no-custom-value case.

Reply:
oh you mean it like that, sorry didn't fully understand your preivous comment

Reply:
i did flip true to false i suppose